django/thunderstore/api/cyberstorm/views/package_preview.py

```diff
@@ -0,0 +1,288 @@
+            filter_q.add(Q(package__name=params["package_identifier"]), Q.AND)
+
+        if not params.get("deprecated", None):
+            exclude_q.add(Q(package__is_deprecated=True), Q.OR)
+
+        if not params.get("nsfw", None):
+            exclude_q.add(Q(has_nsfw_content=True), Q.OR)
+
+        if params.get("q", None):
+            icontains_query = Q()
+            parts = [x for x in params["q"].split(" ") if x]
+
+            for part in parts:
+                for field in (
+                    "package__name",
+                    "package__owner__name",
+                    "package__latest__description",
+                ):
+                    icontains_query &= ~Q(**{f"{field}__icontains": part})
+            exclude_q.add(icontains_query, Q.OR)
+
+        # Add category filters/excludes
+        included = []
+        excluded = []
+        if params.get("section", None):
+            try:
+                section = PackageListingSection.objects.prefetch_related(
+                    "require_categories", "exclude_categories"
+                ).get(slug=params["section"])
+            except PackageListingSection.DoesNotExist:
+                pass
+            else:
+                [included.append(s.slug) for s in section.require_categories.all()]
+                [excluded.append(s.slug) for s in section.exclude_categories.all()]
+
+        if params.get("included_categories", None):
+            included = list(chain(included, params["included_categories"]))
+
+        if params.get("excluded_categories", None):
+            excluded = list(chain(excluded, params["excluded_categories"]))
+
+        if len(included) > 0:
+            included_cats_q = Q()
+            for cat in sorted(set(included)):
+                included_cats_q.add(Q(categories__slug__icontains=cat), Q.OR)
+            filter_q.add(included_cats_q, Q.AND)
+
+        if len(excluded) > 0:
+            excluded_cats_q = Q()
+            for cat in sorted(set(excluded)):
+                excluded_cats_q.add(Q(categories__slug__icontains=cat), Q.OR)
+            exclude_q.add(excluded_cats_q, Q.OR)
+
+        return (
+            PackageListing.objects.active()
+            .filter(
+                review_status__icontained_by=Case(
+                    When(
+                        Q(community__require_package_listing_approval=True),
+                        then=Value("approved"),
+                    ),
+                    default=Value("unreviewed approved"),
+                )
+            )
+            .filter(filter_q)
+            .exclude(exclude_q)
+            .distinct()
+            .prefetch_related(
+                "categories",
+                "community",
+            )
+            .select_related("package__latest", "package__namespace", "package__owner")
+            .annotate(package_total_downloads=Sum("package__versions__downloads"))
+            .annotate(package_total_rating=Count("package__package_ratings"))
```

Comment:
If I remember correctly, this leads to inflated numbers in some scenarios, but I can't recall why that was exactly.

There's a reason it got commented out on the legacy frontend package list view `get_queryset` method here: https://github.com/thunderstore-io/Thunderstore/blob/master/django/thunderstore/repository/views/repository.py#L225

If you want to include the annotations, make sure they get calculated correctly via tests. Make sure those tests ensure packages that are listed in multiple communities and have multiple versions with each having downloads on them yield the correct numbers.

Reply:
Haven't noticed any problems with these numbers, but sure I can make better tests for this 👍